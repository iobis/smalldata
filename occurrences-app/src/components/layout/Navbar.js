/* eslint-disable jsx-a11y/anchor-is-valid */
import classNames from 'classnames'
import React, { useRef, useState } from 'react'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { Link, NavLink } from 'react-router-dom'
import { useOnClickOutside } from '../../hooks/hooks'
import { useTranslation } from 'react-i18next'

export default function Navbar() {
  const { t } = useTranslation()
  const [navbarMenuActive, setNavbarMenuActive] = useState(false)
  const menuRef = useRef()
  const hideNavbarMenu = () => setNavbarMenuActive(false)
  useOnClickOutside(menuRef, hideNavbarMenu)

  return (
    <nav aria-label="main navigation" className="navbar is-info" ref={menuRef} role="navigation">
      <div className="navbar-brand">
        <Link className="navbar-item" onClick={hideNavbarMenu} to="/input-data">
          <p style={{ 'width': 112, 'fontSize': 26, 'fontWeight': 'bold' }}>OBIS</p>
        </Link>
        <a
          className="navbar-burger"
          onClick={() => setNavbarMenuActive(!navbarMenuActive)}
          role="button">
          <span aria-hidden="true"></span>
          <span aria-hidden="true"></span>
          <span aria-hidden="true"></span>
        </a>
      </div>
      <div className={classNames('navbar-menu', { 'is-active': navbarMenuActive })}>
        <div className="navbar-start">
          <NavbarItem onClick={hideNavbarMenu} to="/input-data">
            {t('navbar.inputData')}
          </NavbarItem>
          <NavbarItem onClick={hideNavbarMenu} to="/help">
            {t('navbar.help')}
          </NavbarItem>
        </div>
        <div className="navbar-end">
          <a className="navbar-item" onClick={() => console.log('TBD: logout clicked')}>
            <span className="icon" style={{ 'marginRight': 6 }}><FontAwesomeIcon icon="user"/></span>
            {t('navbar.logout')}
          </a>
        </div>
      </div>
    </nav>
  )
}

const NavbarItem = ({ children, onClick, to }) => (
  <NavLink activeClassName="is-active" className="navbar-item" onClick={onClick} to={to}>
    {children}
  </NavLink>
)
