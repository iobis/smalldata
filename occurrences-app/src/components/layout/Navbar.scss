.navbar {
  &.is-info {
    background-color: transparent;
  }

  .navbar-menu .navbar-start a {
    font-family: var(--title-font);

    &.navbar-item {
      color: var(--font-color);
    }

    &.navbar-item.is-active,
    &.navbar-item:hover {
      color: white;
      background-color: var(--obis-green);
    }
  }

  .navbar-brand {
    p {
      padding-right: 10px;
      font-size: 26px;
      font-weight: bold;
    }

    a.navbar-item.logo {
      background-color: white;
      color: var(--font-color);

      &:hover,
      &:visited {
        background-color: white;
        color: var(--font-color);
      }
    }
  }

  .login-nav-item.navbar-end {
    .navbar-item {
      color: var(--font-color);
      &.auth-button:hover {
        background-color: var(--obis-green);
        color: white;
      }
    }
  }

  .navbar-burger {
    span {
      color: black;
    }
  }
}
