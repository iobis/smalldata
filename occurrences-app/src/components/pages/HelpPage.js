import React from 'react'
import { useTranslation } from 'react-i18next'

export default function HelpPage() {
  const { t } = useTranslation()

  return <div className="helppage"><h3 className="title is-3">  Entering large datasets</h3><p>In case of large datasets being inserted, please <a href={t('helpPage.contactEmail')}>contact
    your node manager</a> for a more efficient method. In case of technical difficulties, please do the same</p></div>
}
