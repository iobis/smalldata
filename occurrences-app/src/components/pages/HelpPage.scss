.help-page {
  h3,
  p {
    text-align: center;
  }
}
