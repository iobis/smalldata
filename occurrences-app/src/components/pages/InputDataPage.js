import Divider from '../layout/Divider'
import PropTypes from 'prop-types'
import React, { useContext, useEffect, useState } from 'react'
import { AuthContext } from '@smalldata/dwca-lib'
import { format } from 'date-fns'
import { getDatasets, getOccurrences } from '@smalldata/dwca-lib/src/clients/SmalldataClient'
import { Link } from 'react-router-dom'
import { useTranslation } from 'react-i18next'

export default function InputDataPage() {
  const { t } = useTranslation()
  const { userRef } = useContext(AuthContext)
  const [occurrences, setOccurrences] = useState([])

  useEffect(() => {
    const fetchOccurrences = async() => {
      const [occurrences, datasets] = await Promise.all([
        getOccurrences({ userRef }),
        getDatasets()
      ])
      const datasetRefToTitle = datasets
        .reduce((acc, dataset) => {
          acc[dataset.id] = dataset.title.value
          return acc
        }, {})
      const occurrencesWithDataset = occurrences
        .map(occurrence => ({
          ...occurrence,
          datasetTitle: datasetRefToTitle[occurrence.dataset]
        }))
      setOccurrences(occurrencesWithDataset)
    }
    fetchOccurrences()
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])

  return (
    <>
      <section className="section">
        <div className="container is-fluid has-text-centered">
          <Link className="button is-info new" to="/input-data/create">
            {t('inputDataPage.inputNewOccurrence')}
          </Link>
        </div>
      </section>
      <Divider>{t('inputDataPage.or')}</Divider>
      <section className="previous-occurrences section">
        <div className="container is-fluid has-text-centered">
          <h4 className="title is-4">{t('inputDataPage.copyPreviousHeader')}</h4>
          <table className="table is-striped is-hoverable is-fullwidth">
            <thead>
              <tr>
                <th/>
                <th>{t('inputDataPage.table.addedAt')}</th>
                <th>{t('inputDataPage.table.scientificName')}</th>
                <th>{t('inputDataPage.table.datasetTitle')}</th>
                <th>{t('inputDataPage.table.occurrenceDate')}</th>
                <th/>
              </tr>
            </thead>
            <tbody>
              {occurrences.map(occurrence => (
                <OccurrenceRow
                  datasetId={occurrence.dataset}
                  key={occurrence.dwcaId}
                  occurrenceDate={occurrence.dwcRecords.occurrence[0].tdwg.eventDate}
                  scientificName={occurrence.dwcRecords.occurrence[0].tdwg.scientificName}
                  {...occurrence}/>
              ))}
            </tbody>
          </table>
        </div>
      </section>
    </>
  )
}

function OccurrenceRow({ addedAtInstant, datasetId, dwcaId, datasetTitle, occurrenceDate, scientificName }) {
  const { t } = useTranslation()
  const addedAtString = addedAtInstant ? format(addedAtInstant, 'D MMMM YYYY') : '—'
  const toCreate = {
    pathname: '/input-data/create',
    state:    { action: 'create', dwcaId, datasetId }
  }
  const toEdit = {
    pathname: '/input-data/update',
    state:    { action: 'update', dwcaId, datasetId }
  }

  return (
    <tr className="occurrence-row">
      <td className="edit">
        <Link className="button is-info" to={toEdit}>
          {t('common.edit')}
        </Link>
      </td>
      <td className="added-at">{addedAtString}</td>
      <td className="scientific-name">{scientificName}</td>
      <td className="dataset-title">{datasetTitle}</td>
      <td className="occurrence-date">{occurrenceDate || '—'}</td>
      <td className="copy">
        <Link className="button is-info copy-previous-entry" to={toCreate}>
          {t('common.copy')}
        </Link>
      </td>
    </tr>
  )
}

OccurrenceRow.propTypes = {
  addedAtInstant: PropTypes.string,
  datasetId:      PropTypes.string.isRequired,
  datasetTitle:   PropTypes.string.isRequired,
  dwcaId:         PropTypes.string.isRequired,
  occurrenceDate: PropTypes.string,
  scientificName: PropTypes.string.isRequired
}
