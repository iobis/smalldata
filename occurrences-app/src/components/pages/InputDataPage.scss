.previous-occurrences {
  margin-top: var(--margin-top);
  .occurrence-row {
    .edit {
      width: 5%;
    }

    .added-at {
      width: 10%;
    }

    .scientific-name {
      width: 10%;
    }

    .dataset-title {
      width: 55%;
    }

    .occurrence-date {
      width: 15%;
    }

    .copy {
      width: 5%;
    }
  }
}

a.button.new-occurrence {
  background-color: var(--obis-green);
  margin-bottom: var(--margin-bottom);

  &:hover {
    background-color: lighten($obis-green, 10%);
  }
}

a.button.is-info.copy-previous-entry {
  background-color: $stepbackground;
}
