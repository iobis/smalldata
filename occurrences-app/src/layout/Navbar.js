import classNames from 'classnames'
import React, { useRef, useState } from 'react'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { Link, NavLink } from 'react-router-dom'
import { useOnClickOutside } from '../hooks/hooks'
import { useTranslation } from 'react-i18next'

export default function Navbar() {
  const { t } = useTranslation()
  const [navbarMenuActive, setNavbarMenuActive] = useState(false)
  const menuRef = useRef()

  const onNavbarItemClick = () => setNavbarMenuActive(false)

  useOnClickOutside(menuRef, () => setNavbarMenuActive(false))

  return (
    <nav className="navbar is-info" role="navigation" ref={menuRef} aria-label="main navigation">
      <div className="navbar-brand">
        <Link to="/input-data" className="navbar-item">
          <p style={{ 'width': 112, 'fontSize': 26, 'fontWeight': 'bold' }}>OBIS</p>
        </Link>
        <a
          role="button"
          className="navbar-burger"
          onClick={() => setNavbarMenuActive(!navbarMenuActive)}>
          <span aria-hidden="true"></span>
          <span aria-hidden="true"></span>
          <span aria-hidden="true"></span>
        </a>
      </div>
      <div className={classNames('navbar-menu', { 'is-active': navbarMenuActive })}>
        <div className="navbar-start">
          <NavbarItem onClick={onNavbarItemClick} to="/input-data">
            {t('navbar.inputData')}
          </NavbarItem>
          <NavbarItem onClick={onNavbarItemClick} to="/help">
            {t('navbar.help')}
          </NavbarItem>
        </div>
        <div className="navbar-end">
          <a className="navbar-item" onClick={() => console.log('TBD: logout clicked')}>
            <span className="icon" style={{ 'marginRight': 6 }}><FontAwesomeIcon icon="user"/></span>
            {t('navbar.logout')}
          </a>
        </div>
      </div>
    </nav>
  )
}

const NavbarItem = ({ children, onClick, to }) => (
  <NavLink activeClassName="is-active" className="navbar-item" to={to} onClick={onClick}>
    {children}
  </NavLink>
)
