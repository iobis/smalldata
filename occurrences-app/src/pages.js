export const INPUT_DATA_PAGE = 'InputDataPage'
export const HELP_PAGE = 'HelpPage'
