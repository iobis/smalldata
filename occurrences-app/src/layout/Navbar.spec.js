import 'jest-dom/extend-expect'
import Navbar from './Navbar'
import React from 'react'
import ReactDom from 'react-dom'
import renderer from 'react-test-renderer'
import TestUtils from 'react-dom/test-utils'
import { INPUT_DATA_PAGE, HELP_PAGE } from '../pages'

describe('Navbar', () => {
  it('renders correctly', () => {
    expect(renderer.create(<Navbar activePage={INPUT_DATA_PAGE}/>).toJSON()).toMatchSnapshot()
    expect(renderer.create(<Navbar activePage={HELP_PAGE}/>).toJSON()).toMatchSnapshot()
  })

  describe('when clicking navbar items', () => {
    let onChange
    let el
    let navbarItems

    beforeAll(() => {
      onChange = jest.fn()
      el = renderDom(<div><Navbar activePage={INPUT_DATA_PAGE} onPageChange={onChange}/></div>)
      navbarItems = el.querySelectorAll('.navbar-start .navbar-item')
    })

    it('renders 2 navbar items', () => {
      expect(navbarItems).toHaveLength(2)
    })

    it('does not call handler', () => {
      expect(onChange).toHaveBeenCalledTimes(0)
    })

    describe('and then clicking first navbar item', () => {
      beforeAll(() => {
        click(navbarItems[0])
      })

      it('calls handler with INPUT_DATA_PAGE', () => {
        expect(onChange).toHaveBeenCalledTimes(1)
        expect(onChange).toHaveBeenNthCalledWith(1, INPUT_DATA_PAGE)
      })

      describe('and then clocking second navbar item', () => {
        beforeAll(() => {
          click(navbarItems[1])
        })

        it('calls handler with HELP_PAGE', () => {
          expect(onChange).toHaveBeenCalledTimes(2)
          expect(onChange).toHaveBeenNthCalledWith(2, HELP_PAGE)
        })
      })
    })
  })
})

function renderDom(Component) {
  return ReactDom.findDOMNode(TestUtils.renderIntoDocument(Component))
}

function click(element) {
  TestUtils.Simulate.click(element)
}
