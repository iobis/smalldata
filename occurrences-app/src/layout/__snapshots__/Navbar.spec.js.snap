// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar renders correctly 1`] = `
<nav
  aria-label="main navigation"
  className="navbar is-info"
  role="navigation"
>
  <div
    className="navbar-brand"
  >
    <a
      className="navbar-item"
    >
      <p
        style={
          Object {
            "fontSize": 26,
            "fontWeight": "bold",
            "width": 112,
          }
        }
      >
        OBIS
      </p>
    </a>
    <a
      className="navbar-burger"
      onClick={[Function]}
      role="button"
    >
      <span
        aria-hidden="true"
      />
      <span
        aria-hidden="true"
      />
      <span
        aria-hidden="true"
      />
    </a>
  </div>
  <div
    className="navbar-menu"
  >
    <div
      className="navbar-start"
    >
      <a
        className="navbar-item is-active"
        onClick={[Function]}
      >
        INPUT DATA
      </a>
      <a
        className="navbar-item"
        onClick={[Function]}
      >
        HELP
      </a>
    </div>
    <div
      className="navbar-end"
    >
      <a
        className="navbar-item"
        onClick={[Function]}
      >
        <span
          className="icon"
          style={
            Object {
              "marginRight": 6,
            }
          }
        >
          <i
            className="fas fa-lg fa-user"
          />
        </span>
        logout
      </a>
    </div>
  </div>
</nav>
`;

exports[`Navbar renders correctly 2`] = `
<nav
  aria-label="main navigation"
  className="navbar is-info"
  role="navigation"
>
  <div
    className="navbar-brand"
  >
    <a
      className="navbar-item"
    >
      <p
        style={
          Object {
            "fontSize": 26,
            "fontWeight": "bold",
            "width": 112,
          }
        }
      >
        OBIS
      </p>
    </a>
    <a
      className="navbar-burger"
      onClick={[Function]}
      role="button"
    >
      <span
        aria-hidden="true"
      />
      <span
        aria-hidden="true"
      />
      <span
        aria-hidden="true"
      />
    </a>
  </div>
  <div
    className="navbar-menu"
  >
    <div
      className="navbar-start"
    >
      <a
        className="navbar-item"
        onClick={[Function]}
      >
        INPUT DATA
      </a>
      <a
        className="navbar-item is-active"
        onClick={[Function]}
      >
        HELP
      </a>
    </div>
    <div
      className="navbar-end"
    >
      <a
        className="navbar-item"
        onClick={[Function]}
      >
        <span
          className="icon"
          style={
            Object {
              "marginRight": 6,
            }
          }
        >
          <i
            className="fas fa-lg fa-user"
          />
        </span>
        logout
      </a>
    </div>
  </div>
</nav>
`;
